Source/DafnyDriver/DafnyNewCli.cs

```diff
@@ -257,30 +260,56 @@ public static class DafnyNewCli {
     return builder;
   }
 
-  private static async Task<DafnyFile?> HandleDafnyProject(DafnyOptions options,
-    IFileSystem fileSystem, ErrorReporter reporter,
+  private static async IAsyncEnumerable<DafnyFile> HandleDafnyProject(IFileSystem fileSystem, DafnyOptions options,
+    ErrorReporter reporter,
     Uri uri,
-    IToken origin,
+    IToken uriOrigin,
     bool asLibrary) {
     if (!asLibrary) {
-      reporter.Error(MessageSource.Project, origin, "Using a Dafny project file as a source file is not supported.");
-      return null;
+      reporter.Error(MessageSource.Project, uriOrigin, "Using a Dafny project file as a source file is not supported.");
+      yield break;
     }
 
-    var outputWriter = new StringWriter();
-    var errorWriter = new StringWriter();
-    var exitCode = await DafnyNewCli.Execute(new WritersConsole(TextReader.Null, outputWriter, errorWriter),
-      new[] { "build", "-t=lib", uri.LocalPath, "--verbose" });
-    if (exitCode != 0) {
-      var output = outputWriter + errorWriter.ToString();
-      reporter.Error(MessageSource.Project, origin,
-        $"Could not build a Dafny library from {uri.LocalPath} because:\n{output}");
-      return null;
+    var dependencyProject = await DafnyProject.Open(fileSystem, options, uri, uriOrigin);
+    var dependencyOptions =
+      DooFile.CheckAndGetLibraryOptions(reporter, uri, options, uriOrigin, dependencyProject.Options,
+        new Dictionary<Option, OptionCompatibility.OptionCheck> {
+          { CommonOptionBag.Libraries, OptionCompatibility.NoOpOptionCheck }
+        });
+    if (dependencyOptions != null) {
+      if (options.Get(DafnyFile.UnsafeDependencies) || !options.Verify) {
+        foreach (var libraryRootSetFile in dependencyProject.GetRootSourceUris(fileSystem)) {
+          var file = DafnyFile.HandleDafnyFile(fileSystem, reporter, dependencyOptions, libraryRootSetFile,
+            dependencyProject.StartingToken, true, false);
+          if (file != null) {
+            yield return file;
+          }
+        }
+      } else {
+        if (options.Verbose) {
+          await options.OutputWriter.WriteLineAsync($"Building dependency {options.GetPrintPath(uri.LocalPath)}");
+        }
+
+        dependencyOptions.Compile = true;
+        dependencyOptions.RunAfterCompile = false;
+        var libraryBackend = new LibraryBackend(dependencyOptions);
+        dependencyOptions.CompilerName = libraryBackend.TargetId;
+
+        dependencyOptions.DafnyProject = dependencyProject;
+        dependencyOptions.CliRootSourceUris.Clear();
+        dependencyOptions.Compile = true;
+        dependencyOptions.RunAfterCompile = false;
+        var exitCode = await SynchronousCliCompilation.Run(dependencyOptions);
```

Comment:
How confident are we that we've eliminated all the global state that might leak to this in-memory recursive call to the pipeline? I wonder if you want to leave things with the subprocess call for now, to avoid introducing extra risk with this PR.

Reply:
I'd rather not to the subprocess call because both setting up the arguments and processing the result is awkward. I think a bug in this code is most likely to cause the whole process to fail. I'm not worried about producing a doo file that can be consumed but has not been validated (unsoundness).